### PR TITLE
Enable loading and saving quantized decode checkpoint.

### PR DESCRIPTION
**Pull request by singh-mitali**
Adding support to save and load decode params quantized on the fly using AQT.

### CODE REVIEW

**Comment by vipannalla** [comment]
Looks good to me at high level. Can you also paste the mem_stats before, during, and after loading from quantized checkpoint and from unquantized checkpoint?

**Comment by mitalisi** [comment]
> Looks good to me at high level. Can you also paste the mem_stats before, during, and after loading from quantized checkpoint and from unquantized checkpoint?

Loading quantized 7b checkpoint
python MaxText/decode.py MaxText/configs/base.yml tokenizer_path=assets/tokenizer.llama2 load_parameters_path=gs://msingh-bkt/checkpoints/quantized/aqt/llama2-7b-chat max_prefill_predict_length=1024 max_target_length=2048 model_name=llama2-7b ici_fsdp_parallelism=1 ici_autoregressive_parallelism=1 ici_tensor_parallelism=-1 scan_layers=false weight_dtype=bfloat16 per_device_batch_size=11 quantization=int8  quantize_on_fly=false

Memstats: After load_params:
        Using 836.1 MiB / 15.7 GiB (5.184923%) on TPU_0(process=0,(0,0,0,0))
        Using 836.0 MiB / 15.7 GiB (5.184338%) on TPU_1(process=0,(1,0,0,0))
        Using 836.0 MiB / 15.7 GiB (5.184338%) on TPU_2(process=0,(2,0,0,0))
        Using 836.0 MiB / 15.7 GiB (5.184338%) on TPU_3(process=0,(3,0,0,0))
        Using 836.0 MiB / 15.7 GiB (5.184338%) on TPU_4(process=0,(0,1,0,0))
        Using 836.0 MiB / 15.7 GiB (5.184338%) on TPU_5(process=0,(1,1,0,0))
        Using 836.0 MiB / 15.7 GiB (5.184338%) on TPU_6(process=0,(2,1,0,0))
        Using 836.0 MiB / 15.7 GiB (5.184338%) on TPU_7(process=0,(3,1,0,0))



Loading unquantized 7b checkpoint
python MaxText/decode.py MaxText/configs/base.yml tokenizer_path=assets/tokenizer.llama2 load_parameters_path=gs://inference-benchmarks/models/llama2-7b-chat/2024-05-24-12-39/param-only-decode-ckpt-maxtext/checkpoints/0/items max_prefill_predict_length=1024 max_target_length=2048 model_name=llama2-7b ici_fsdp_parallelism=1 ici_autoregressive_parallelism=1 ici_tensor_parallelism=-1 scan_layers=false weight_dtype=bfloat16 per_device_batch_size=11

Memstats: After load_params:
        Using 1.6 GiB / 15.7 GiB (9.963308%) on TPU_0(process=0,(0,0,0,0))
        Using 1.6 GiB / 15.7 GiB (9.962724%) on TPU_1(process=0,(1,0,0,0))
        Using 1.6 GiB / 15.7 GiB (9.962724%) on TPU_2(process=0,(2,0,0,0))
        Using 1.6 GiB / 15.7 GiB (9.962724%) on TPU_3(process=0,(3,0,0,0))
        Using 1.6 GiB / 15.7 GiB (9.962724%) on TPU_4(process=0,(0,1,0,0))
        Using 1.6 GiB / 15.7 GiB (9.962724%) on TPU_5(process=0,(1,1,0,0))
        Using 1.6 GiB / 15.7 GiB (9.962724%) on TPU_6(process=0,(2,1,0,0))
        Using 1.6 GiB / 15.7 GiB (9.962724%) on TPU_7(process=0,(3,1,0,0))

**Comment by vipannalla** [comment]
Thanks, lgtm.

**Review by vipannalla — APPROVED** [review]

File: MaxText/max_utils.py [hunk 1]
```diff
@@ -439,22 +440,47 @@ def init_initial_state(model, tx, config, is_training, key):
...
   if is_training:
     return init_training_state(model.apply, model_vars, tx)
-  return init_decode_state(model.apply, model_vars)
+  return init_decode_state(model_vars)
 
 
-def load_decode_model_vars(model, config, rng, mesh):
-  state, _ = setup_decode_state(model, config, rng, mesh, None)
-  return state.params
+def setup_decode_state(model, config, rng, mesh):
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/max_utils.py:446]: We use decode for all inference related state in this file.

File: end_to_end/tpu/test_decode_load_quantized_ckpt.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply end_to_end/tpu/test_decode_load_quantized_ckpt.sh:1]: Yes we can do so.

File: MaxText/max_utils.py [hunk 3]
```diff
@@ -411,9 +412,9 @@ def unbox_logicallypartioned(boxed_pytree):
   )
 
 
-def init_decode_state(apply_fn, params):
+def init_decode_state(params):
   """Init train state with null opt state for decode."""
-  state = train_state.TrainState(step=0, apply_fn=apply_fn, params=params, tx=None, opt_state={})  # type: ignore
+  state = train_state.TrainState(step=0, apply_fn=None, params=params, tx=None, opt_state={})  # type: ignore
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/max_utils.py:417]: No - its the init function as discussed offline.

File: MaxText/layers/quantizations.py [hunk 4]
```diff
@@ -45,21 +46,43 @@ def dot_general_cls(self):
...
 
+def _rhs_axis_metadata_wrapper(x: jnp.ndarray, no_sharding_axis: Sequence[int], mesh_axes: Tuple[str, ...], is_tiled: bool):
+  mesh_axes = list(mesh_axes)
+  if is_tiled:
+    # The original rank of x will be doubled, e.x. [a, b, c] --> [_, a', _, b', _, c'].
+    # The mesh axes should be doubled too.
+    assert len(x.shape) >= 2 * len(mesh_axes)
+    mesh_axes = [None] * (len(x.shape) - 2 * len(mesh_axes)) + sum(
+        ([None, axis] for axis in mesh_axes), []
+    )
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/layers/quantizations.py:57]: Its for subchannel quant which we will also checkin this week. IMO its ok to have a more generic function checked in as far as we are calling it with the right params.

File: MaxText/maxengine.py [hunk 5]
```diff
@@ -71,57 +71,62 @@ def __init__(self, config):
...
     self.kv_cache_shardings = None
-    self.state_mesh_annotations = None
 
   def load_params(self, *args, **kwargs) -> Params:
     """Load Parameters, typically from GCS"""
     # pylint: disable=unused-argument
-    state, self.state_mesh_annotations = max_utils.setup_decode_state(self.model, self.config, self.rng, self._mesh, None)
+
+    if self.model.quant and not self.config.quantize_on_fly:
+      print("Loading from the quantized checkpoint...")
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/maxengine.py:85]: No. Setting quantize_on_fly =false and quantization=int8 implied that we are loading an int8 checkpoint.

[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/maxengine.py:85]: Oops, I mean `quantize_on_fly = True`  and `quantization=int8` generate random int8  weights?

[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/maxengine.py:85]: Yes.

File: end_to_end/tpu/test_decode_load_quantized_ckpt.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+dry_run=false
+model='llama2-7b'
+
+while getopts "nm:" opt
```
Comment (gobbleturk) [inline end_to_end/tpu/test_decode_load_quantized_ckpt.sh:6]: I didn't realize we already had this pattern of bash inputs in MaxText tests... I would prefer to be consistent with the majority of other tests but this is also fine

**Review by gobbleturk — APPROVED** [review]
Looks great, thank you! 

I requested some minior variable name changes / comment additions

File: MaxText/layers/quantizations.py [hunk 7]
```diff
@@ -40,26 +41,48 @@
 class Quantization:
   """Base class for quantization configurations"""
 
-  def dot_general_cls(self):
+  def dot_general_cls(self, mesh_axes: Tuple[str, ...] = ()):
     """Placeholder for dot_general implementation in subclasses."""
     pass
 
 
+def _rhs_axis_metadata_wrapper(x: jnp.ndarray, no_sharding_axis: Sequence[int], mesh_axes: Tuple[str, ...], is_tiled: bool):
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/layers/quantizations.py:49]: Done

**Comment by mitalisi** [comment]
> Looks great, thank you!
> 
> I requested some minior variable name changes / comment additions

Thanks for the review - made the changes.

File: end_to_end/tpu/test_decode.sh [hunk 8]
```diff
@@ -1,36 +1,12 @@
 #!/bin/bash
 set -ex
 
-NUM_TOKEN_THRESHOLD=${1}
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply end_to_end/tpu/test_decode.sh:4]: Will do.

File: MaxText/layers/quantizations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/layers/quantizations.py]: Done

File: MaxText/generate_param_only_checkpoint.py [hunk 9]
```diff
@@ -92,7 +92,8 @@ def _save_decode_checkpoint(config, state, checkpoint_manager):
   """Generate checkpoint for decode from the training_state."""
   with jax.spmd_mode("allow_all"):
     decode_state = max_utils.init_decode_state(
-      None, jax.tree_util.tree_map(lambda x: x.astype(jax.numpy.bfloat16), state.params))
+      jax.tree_util.tree_map(lambda x: x.astype(jax.numpy.bfloat16), state.params)
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/generate_param_only_checkpoint.py:95]: Ack

File: MaxText/maxengine.py [hunk 10]
```diff
@@ -71,57 +71,62 @@ def __init__(self, config):
     self.model = models.Transformer(config, mesh=self._mesh, quant=quant)
     self.replicated_sharding = jax.sharding.NamedSharding(self._mesh, P(None))
 
+    self.state_mesh_annotations = None
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/maxengine.py:74]: Reverted.

**Pull request closed without merging**